### PR TITLE
fix(pwa): avoid caching large files

### PR DESCRIPTION
**Pull request by MrMurdock11**
## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (refactoring and improving code)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Description
I'm going to try to solve an issue related to caching files that are chunked for transfer (usually they have a 206 status) that we have discussed here GH-1651. To avoid this error, we should only cache files that have a response code of 200.

image

### CODE REVIEW

**Review by cotes2020 — CHANGES_REQUESTED** [review]
1. Please change the scop of the commit message to `pwa`:
   `fix(cache):` -> `fix(pwa):`.

2. This change doesn‘t seem to work. It breaks PWA's ability to respond offline to cached resources (e.g. JS/CSS files of type `cors`). You can verify this with local testing. `JEKYLL_ENV=production bundle exec jekyll b -l`

   Visit http://localhost:4000/, open a few random posts, then disconnect from the network and visit the same history again:
   
   Screenshot 2024-04-14 at 21 26 05

File: assets/js/pwa/sw.js [hunk 1]
```diff
@@ -84,7 +84,7 @@ self.addEventListener('fetch', (event) => {
       return fetch(event.request).then((response) => {
         const url = event.request.url;
 
-        if (purge || event.request.method !== 'GET' || !verifyUrl(url)) {
+        if (purge || response.status !== 200 || event.request.method !== 'GET' || !verifyUrl(url)) {
```
Comment (cotes2020) [inline assets/js/pwa/sw.js:87]: Changing the condition to hit `206` to return the response directly would restore the functionality of the PWA, but I'm not sure that's the best way to do it

```suggestion
        if (purge || response.status == 206 || event.request.method !== 'GET' || !verifyUrl(url)) {
```

Comment (MrMurdock11) [reply assets/js/pwa/sw.js:87]: any thoughts?

**Comment by MrMurdock11** [comment]
@cotes2020 What do you think if I fix the 206 code to filter the request and finalize this PR, but create a new discussion to find a better solution?

**Comment by kungfux** [comment]
> @cotes2020 What do you think if I fix the 206 code to filter the request and finalize this PR, but create a new discussion to find a better solution?

I didn't elaborate on this much, but I have seen the problem in action as it causes errors for newly developed features and so should be fixed before release, I think.

Is there anything else we should fix except 206? I mean maybe it's a good idea to fix what we have discovered so far only?

**Comment by cotes2020** [comment]
The comments I left [above](https://github.com/cotes2020/jekyll-theme-chirpy/pull/1672#discussion_r1564730581) were mainly to get your opinions on whether there is a better solution.

Unless you have a better way, detecting whether the response status code is 206 is an acceptable solution. 

> If innerResponse’s [status](https://fetch.spec.whatwg.org/#concept-response-status) is 206, return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a TypeError.
>
> Ref: <https://w3c.github.io/ServiceWorker/#cache-put>

**Comment by MrMurdock11** [comment]
@cotes2020 added the changes to your comment

**Comment by cotes2020** [comment]
Sorry guys, I made a mistake while reviewing, returning directly after `event.request.headers.has('range')` is met would have the undesirable result that `fontwesome` styles could not be cached.

So, I'm going to rewrite this commit to fix this issue.

**Pull request merged**